x-pack/plugins/ml/public/url_generator.ts

```diff
@@ -64,39 +208,265 @@ export class MlUrlGenerator implements UrlGeneratorsDefinition<typeof ML_APP_URL
+      mlTimeSeriesExplorer.detectorIndex = detectorIndex;
+    }
+    if (entities !== undefined) {
+      mlTimeSeriesExplorer.entities = entities;
+    }
+    appState.mlTimeSeriesExplorer = mlTimeSeriesExplorer;
+
+    if (zoom) appState.zoom = zoom;
+    if (query)
+      appState.query = {
+        query_string: query,
+      };
+    url = setStateToKbnUrl<TimeSeriesExplorerGlobalState>(
+      '_g',
+      queryState,
+      { useHash: false, storeInHashQuery: false },
+      url
+    );
+    url = setStateToKbnUrl<TimeSeriesExplorerAppState>(
+      '_a',
+      appState,
+      { useHash: false, storeInHashQuery: false },
+      url
+    );
+
+    return url;
+  }
+
+  /**
+   * Creates URL to the DataFrameAnalytics Exploration page
+   */
+  private createDataframeAnalyticsExplorationUrl(
+    mlUrlGeneratorState: DataFrameAnalyticsExplorationUrlState
+  ): string {
+    let url = `${this.params.appBasePath}/${ML_PAGES.DATA_FRAME_ANALYTICS_EXPLORATION}`;
+    const { params } = MlUrlGenerator.extractParams<DataFrameAnalyticsExplorationUrlState>(
+      mlUrlGeneratorState
+    );
+
+    if (!isEmpty(params)) {
+      const { jobId, analysisType } = params;
+      const queryState: DataFrameAnalyticsExplorationQueryState = {
+        ml: {
+          jobId,
+          analysisType,
+        },
+      };
+
+      url = setStateToKbnUrl<DataFrameAnalyticsExplorationQueryState>(
+        '_g',
+        queryState,
+        { useHash: false, storeInHashQuery: false },
+        url
+      );
+    }
+
+    return url;
+  }
 
-    let url = `${this.params.appBasePath}/explorer`;
-    url = setStateToKbnUrl<ExplorerQueryState>('_g', queryState, { useHash: false }, url);
-    url = setStateToKbnUrl('_a', appState, { useHash: false }, url);
+  /**
+   * Creates URL to the DataFrameAnalytics page
+   */
+  private createDataframeAnalyticsUrl(mlUrlGeneratorState: DataFrameAnalyticsUrlState): string {
+    let url = `${this.params.appBasePath}/${ML_PAGES.DATA_FRAME_ANALYTICS}`;
+    const { params } = MlUrlGenerator.extractParams<DataFrameAnalyticsUrlState>(
+      mlUrlGeneratorState
+    );
+
+    if (!isEmpty(params)) {
+      const { jobId, groupIds } = params;
+      const queryState: Partial<DataFrameAnalyticsQueryState> = {
+        jobId,
+        groupIds,
+      };
+
+      url = setStateToKbnUrl<Partial<DataFrameAnalyticsQueryState>>(
+        'mlManagement',
+        queryState,
+        { useHash: false, storeInHashQuery: false },
+        url
+      );
+    }
+
+    return url;
+  }
 
+  /**
+   * Creates URL to the Data Visualizer page
+   */
+  private createDataVisualizerUrl({ page }: DataVisualizerUrlState): string {
+    return `${this.params.appBasePath}/${page}`;
+  }
+
+  /**
+   * Creates generic index based search ML url
+   * e.g. `jobs/new_job/datavisualizer?index=3da93760-e0af-11ea-9ad3-3bcfc330e42a`
+   */
+  private createIndexBasedMlUrl(mlGenericUrlState: MlGenericUrlState): string {
```

Comment:
This should have `dataVisualizer` in the function name to clarify the pages it's building URLs for. `createIndexDataVisualizerUrl` or similar?

Reply:
Updated here https://github.com/elastic/kibana/pull/75696/commits/49e56edc23a2a9ddc12e9145add11bba8f2b2e7f